### ISSUE TITLE
[video_player][ios] IOSurface stays in the memory even after VideoPlayerController was disposed

### ISSUE THREAD

**Issue by bs-vasyl-trs** [flutter/flutter#86477]
I faced the problem on iOS that even after disposing VideoPlayerController and removing VideoPlayer from widgets tree the IOSurface stays in the memory and as a result - general memory usage is increased, so doing some heavy computations even with removed video player might lead to memory presure warning or even app termination due to memory per-process-limit reason.
I was using Xcode memory tab to check it and Allocation & Leaks instrument
![allocation](https://raw.githubusercontent.com/bs-vasyl-trs/flutter_video_memory_issue/main/allocation_details.png)
![xcode](https://raw.githubusercontent.com/bs-vasyl-trs/flutter_video_memory_issue/main/xcode_memory.png)

Please have a look into this [example](https://github.com/bs-vasyl-trs/flutter_video_memory_issue)
To reproduce the issue:

1. open the project with Xcode & run it
2. open memory tab 
3. try to press '+' icon to add player to the view and after that remove it with '-' icon (you can repeat it several times)
4. check that memory is not released and memory usage basically on the same level as it jumped when the '+' was pressed for the first time

**Target Platform:** iOS
**Target OS version/browser:** 14.4.2
**Devices:** iPhone 8 physical device

## Logs

<details>
<summary>Logs</summary>

```
`flutter analyze`
Analyzing video_player_memory_issue...                                  
No issues found! (ran in 2.2s)
```

```
`flutter doctor -v`
[✓] Flutter (Channel master, 2.4.0-5.0.pre.55, on Mac OS X 10.15.7 19H1217 darwin-x64, locale en-UA)
    • Flutter version 2.4.0-5.0.pre.55 at /Users/vasylterletskyi/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b549e829e8 (13 hours ago), 2021-07-14 15:51:10 -0700
    • Engine revision 093c4ed1a3
    • Dart version 2.14.0 (build 2.14.0-312.0.dev)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.58.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0
```

</details>

**Comment by maheshmnj** [comment]
Hi @bs-vasyl-trs,
Thanks for filing the issue I am able to reproduce the issue on the latest stable channel but on the master channel I don't see such issue, I used the latest video player plugin ```video_player: ^2.1.10```. I would request you to upgrade to the latest master channel by running. 

```
flutter channel master
flutter upgrade
``` 
and see  if you are able to reproduce the issue on the latest master channel ``` master, 2.4.0-5.0.pre.87``` too.

<details>
<summary>Output</summary>

|    |  Video Player Memory Usage  | 
| :---: | :---: | 
| Stable Channel |  <img src="https://user-images.githubusercontent.com/31410839/126176318-c6ebc159-b01b-48f6-8793-0ccf3f6d6d0f.jpg"  />	|
| Mater Channel | <img src="https://user-images.githubusercontent.com/31410839/126178019-9e061c6e-5321-4646-8575-f89c58ba3438.png"  /> |

</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (2 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```

```
[✓] Flutter (Channel master, 2.4.0-5.0.pre.87, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.4.0-5.0.pre.87 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e084e9d946 (2 days ago), 2021-07-17 12:00:21 -0700
    • Engine revision 14a9e9a50e
    • Dart version 2.14.0 (build 2.14.0-321.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • 6A434744-B348-4FB3-AB10-131292174456 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```
</details>

Thank you.

**Comment by bs-vasyl-trs** [comment]
Hi @maheshmnj,
thanks for quick response. I've tried with the latest version in master `2.4.0-5.0.pre.90` and issue still persists for real device - iPhone 8, iOS 14.4
![Screenshot 2021-07-19 at 9 29 18 PM](https://user-images.githubusercontent.com/65649897/126216894-03ae5f92-1aca-4954-9eb4-09e0c38f87b8.png)
I also tried on Simulators (12 Max Pro & 8) and can confirm that issue not there, but I think simulators are not representable for tracking memory issues as it's managed by the MacOS, furthermore, instances of IOSurface which stay in the memory on real devices are not even allocated for simulators
![Screenshot 2021-07-19 at 9 59 58 PM](https://user-images.githubusercontent.com/65649897/126217397-62bc5a13-3d27-4d22-a194-728cee7e143c.png)
Could you please try to reproduce on the real device. Thanks

**Comment by maheshmnj** [comment]
Hi @bs-vasyl-trs,
My Colleague @TahaTesser verified it on the latest master channel with a real IOS device he was able to reproduce the issue. 

<img width="1678" alt="Screenshot 2021-07-20 at 10 24 31 AM" src="https://user-images.githubusercontent.com/31410839/126280499-ed39b71e-4a2a-4d73-901b-599352f6e47f.png">

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.4.0-5.0.pre.105, on macOS 12.0 21A5284e darwin-x64, locale en-GB)
    • Flutter version 2.4.0-5.0.pre.105 at /Users/tahatesser/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a5b9b5774a (4 hours ago), 2021-07-19 20:23:45 -0700
    • Engine revision 9edde7423f
    • Dart version 2.14.0 (build 2.14.0-332.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/tahatesser/Code/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/SDK
    • Java binary at: /Users/tahatesser/Code/openjdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_292-b10)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.58.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (3 available)
    • Taha's iPhone (mobile) • 00008020-001059882212002E • ios            • iOS 14.6 18F72
    • macOS (desktop)        • macos                     • darwin-x64     • macOS 12.0 21A5284e darwin-x64
    • Chrome (web)           • chrome                    • web-javascript • Google Chrome 91.0.4472.164
! Doctor found issues in 1 category.
```
</details>


Thank you.

**Comment by ColdPaleLight** [comment]
I reproduced the issue，I found that most of the IOSurface stay in memory because they are held by CVMetalTextureCache.
![image](https://user-images.githubusercontent.com/31977171/129311975-80bc3196-e5f4-4f62-b1da-53c3825d3d76.png)
So I try to flush currently unused resources when the FlutterDarwinExternalTextureMetal is released. 
https://github.com/flutter/engine/pull/28059

**Before changes:**
When VideoPlayer is removed, many IOSurfaces generated by VideoPlayer remain in the memory. The memory use is 124M

**After changes:**
When VideoPlayer is removed, only one IOSurface generated by VideoPlayer remains in the memory, the current memory is 89.1M
![image](https://user-images.githubusercontent.com/31977171/129315365-8bf308e2-d8db-45dc-90ee-fed02831ee2d.png)
